File: frontend/pms_web/src/lib/repositories/reports.ts
import apiClient from "@/lib/api/client";

export interface SalesSummary {
  currency: string;
  start_date: string;
  end_date: string;
  sale_count: number;
  gross_total_cents: number;
  discount_cents: number;
  credit_total_cents: number;
  refund_total_cents: number;
  net_sales_cents: number;
  by_payment_method: { method: string; count: number; total_cents: number }[];
}

export interface DailySalesPoint {
  day: string;
  sale_count: number;
  total_cents: number;
  credit_cents: number;
  refund_cents: number;
  net_sales_cents: number;
}

export interface TopDrug {
  drug_id: string;
  drug_name: string;
  units_sold: number;
  revenue_cents: number;
}

export interface DashboardOverview {
  sales_summary: SalesSummary;
  stock_valuation: {
    cost_value_cents: number;
    sell_value_cents: number;
    total_units: number;
  };
  low_stock_count: number;
  expiring_soon_count: number;
  pending_prescriptions: number;
}

const range = (start?: string, end?: string, signal?: AbortSignal) => ({
  params: { start, end },
  signal,
});

export const reportsRepo = {
  overview: async (): Promise<DashboardOverview> => {
    const { data } = await apiClient.get("/v1/reports/overview");
    return data;
  },
  salesSummary: async (
    start?: string,
    end?: string,
    signal?: AbortSignal,
  ): Promise<SalesSummary> => {
    const { data } = await apiClient.get(
      "/v1/reports/sales-summary",
      range(start, end, signal),
    );
    return data;
  },
  salesDaily: async (
    start?: string,
    end?: string,
    signal?: AbortSignal,
  ): Promise<DailySalesPoint[]> => {
    const { data } = await apiClient.get(
      "/v1/reports/sales-daily",
      range(start, end, signal),
    );
    return data.items;
  },
  topDrugs: async (
    start?: string,
    end?: string,
    limit = 10,
    signal?: AbortSignal,
  ): Promise<TopDrug[]> => {
    const { data } = await apiClient.get("/v1/reports/top-drugs", {
      params: { start, end, limit },
      signal,
    });
    return data.items;
  },
  stockValuation: async () => {
    const { data } = await apiClient.get("/v1/reports/stock-valuation");
    return data as {
      cost_value_cents: number;
      sell_value_cents: number;
      total_units: number;
    };
  },
};

File: backend/services/pms_service/app/schemas/reports.py
from __future__ import annotations

from pydantic import BaseModel


class PaymentMethodBreakdown(BaseModel):
    method: str
    count: int
    total_cents: int


class SalesSummary(BaseModel):
    currency: str = "GHS"
    start_date: str
    end_date: str
    sale_count: int
    gross_total_cents: int
    discount_cents: int
    credit_total_cents: int = 0
    refund_total_cents: int = 0
    net_sales_cents: int = 0
    by_payment_method: list[PaymentMethodBreakdown]


class DailySalesPoint(BaseModel):
    day: str
    sale_count: int
    total_cents: int
    credit_cents: int = 0
    refund_cents: int = 0
    net_sales_cents: int = 0


class DailySalesSeries(BaseModel):
    items: list[DailySalesPoint]


class TopDrug(BaseModel):
    drug_id: str
    drug_name: str
    units_sold: int
    revenue_cents: int


class TopDrugList(BaseModel):
    items: list[TopDrug]


class CountByKey(BaseModel):
    key: str
    count: int


class DispensingVolume(BaseModel):
    by_status: list[dict]
    by_source: list[dict]


class MovementBreakdownItem(BaseModel):
    reason: str
    movement_count: int
    net_units: int


class MovementBreakdownList(BaseModel):
    items: list[MovementBreakdownItem]


class StockValuation(BaseModel):
    cost_value_cents: int
    sell_value_cents: int
    total_units: int


class DashboardOverview(BaseModel):
    sales_summary: SalesSummary
    stock_valuation: StockValuation
    low_stock_count: int
    expiring_soon_count: int
    pending_prescriptions: int
